File: integ/server/gws/drive/query.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { docPlainText } from '../docs/body.ts'
import type { GwsState } from '../store/state.ts'
import type { DriveItem } from '../store/types.ts'

export interface QueryClause {
  field: string
  op: string
  value: string
}

export function unescapeQ(value: string): string {
  let out = ''
  for (let i = 0; i < value.length; i += 1) {
    if (value[i] === '\\' && i + 1 < value.length) {
      i += 1
      out += value[i]
      continue
    }
    out += value[i]
  }
  return out
}

// AND-only Drive query parser covering the clauses mirage and the gws
// commands emit: 'id' in parents, name = / contains, mimeType =, trashed,
// modifiedTime >= / <.
export function parseDriveQuery(q: string): QueryClause[] {
  const clauses: QueryClause[] = []
  let depth = false
  let current = ''
  const parts: string[] = []
  for (let i = 0; i < q.length; i += 1) {
    const c = q[i] as string
    // Drive escapes a quote inside a quoted value as \', so the splitter
    // has to step over the pair the way the clause regexes below already
    // do; toggling on it would end the value early and swallow the ` and `
    // that follows into the same clause.
    if (depth && c === '\\' && i + 1 < q.length) {
      current += c + (q[i + 1] as string)
      i += 1
      continue
    }
    if (c === "'") depth = !depth
    if (!depth && q.slice(i, i + 5) === ' and ') {
      parts.push(current)
      current = ''
      i += 4
      continue
    }
    current += c
  }
  if (current.trim() !== '') parts.push(current)
  for (const raw of parts) {
    const part = raw.trim()
    let m = /^'((?:[^'\\]|\\.)*)'\s+in\s+parents$/.exec(part)
    if (m !== null) {
      clauses.push({ field: 'parents', op: 'in', value: unescapeQ(m[1] as string) })
      continue
    }
    m = /^(\w+)\s*(=|!=|>=|<=|>|<|contains)\s*'((?:[^'\\]|\\.)*)'$/.exec(part)
    if (m !== null) {
      clauses.push({ field: m[1] as string, op: m[2] as string, value: unescapeQ(m[3] as string) })
      continue
    }
    m = /^(\w+)\s*=\s*(true|false)$/.exec(part)
    if (m !== null) {
      clauses.push({ field: m[1] as string, op: '=', value: m[2] as string })
      continue
    }
    throw new Error(`unsupported query clause: ${part}`)
  }
  return clauses
}

// Everything the live index searches for `fullText`: the display name, a
// Doc's text across every tab, a Sheet's cell values, and an uploaded file's bytes.
// Case-insensitive, the way the real search index answers.
export function fullTextOf(st: GwsState, item: DriveItem): string {
  const parts: string[] = [item.name]
  const doc = st.docs.get(item.id)
  if (doc !== undefined) parts.push(docPlainText(doc))
  const sheet = st.sheets.get(item.id)
  if (sheet !== undefined) {
    for (const tab of sheet.tabs) parts.push([...tab.cells.values()].join(' '))
  }
  if (item.content.length > 0) parts.push(item.content.toString('utf8'))
  return parts.join('\n')
}

export function matchClause(st: GwsState, item: DriveItem, clause: QueryClause): boolean {
  switch (clause.field) {
    case 'parents':
      return item.parents.includes(clause.value)
    case 'name':
      if (clause.op === 'contains') return item.name.includes(clause.value)
      if (clause.op === '!=') return item.name !== clause.value
      return item.name === clause.value
    case 'mimeType':
      if (clause.op === 'contains') return item.mimeType.includes(clause.value)
      if (clause.op === '!=') return item.mimeType !== clause.value
      return item.mimeType === clause.value
    case 'fullText':
      // The live API defines only `contains` for fullText; any other
      // operator is an invalid query, reported as the 400 the caller
      // catches below.
      if (clause.op !== 'contains') {
        throw new Error(`unsupported operator for fullText: ${clause.op}`)
      }
      return fullTextOf(st, item).toLowerCase().includes(clause.value.toLowerCase())
    case 'trashed':
      return item.trashed === (clause.value === 'true')
    case 'modifiedTime': {
      if (clause.op === '>=') return item.modifiedTime >= clause.value
      if (clause.op === '<') return item.modifiedTime < clause.value
      if (clause.op === '>') return item.modifiedTime > clause.value
      if (clause.op === '<=') return item.modifiedTime <= clause.value
      return item.modifiedTime === clause.value
    }
    default:
      throw new Error(`unsupported query field: ${clause.field}`)
  }
}

File: python/mirage/core/qdrant/query.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
import logging
import uuid
from collections.abc import Callable
from typing import Any

from qdrant_client import models
from qdrant_client.http.exceptions import UnexpectedResponse

from mirage.accessor.qdrant import QdrantAccessor
from mirage.core.qdrant.naming import group_name, row_stem
from mirage.core.qdrant.payload import field_value
from mirage.core.render.json import value_text

logger = logging.getLogger(__name__)

SCROLL_BATCH = 256


def _json_scalar(text: str) -> bool | int | float | None:
    """The non-string JSON scalar a rendered group segment also spells.

    A group value renders through ``value_text``, so a boolean or a number
    lists as its compact JSON and the segment alone cannot say which type
    the payload holds. Only a spelling ``value_text`` would produce counts:
    ``007``, ``-0`` and ``1.50`` are strings and nothing else.

    Args:
        text (str): the decoded group segment.
    """
    try:
        parsed = json.loads(text)
    except ValueError:
        return None
    if isinstance(parsed, bool):
        return parsed
    if isinstance(parsed, (int, float)) and value_text(parsed) == text:
        return parsed
    return None


def _condition(column: str,
               text: str) -> models.FieldCondition | models.Filter:
    """What one rendered group segment matches in the payload.

    The string itself always, and when the segment also spells a JSON
    scalar, that typed value too, so descending into the ``true`` or
    ``1.5`` directory the listing advertised finds the boolean or float
    points behind it. A number matches as a closed range, which Qdrant
    applies to integer and float payloads alike where ``match`` does not.

    Args:
        column (str): the payload field the group level is named from.
        text (str): the decoded group segment.
    """
    as_text = models.FieldCondition(key=column,
                                    match=models.MatchValue(value=text))
    scalar = _json_scalar(text)
    if scalar is None:
        return as_text
    if isinstance(scalar, bool):
        typed = models.FieldCondition(key=column,
                                      match=models.MatchValue(value=scalar))
    else:
        typed = models.FieldCondition(key=column,
                                      range=models.Range(gte=scalar,
                                                         lte=scalar))
    return models.Filter(should=[as_text, typed])


def _filter(filters: dict[str, str]) -> models.Filter | None:
    if not filters:
        return None
    return models.Filter(
        must=[_condition(column, value) for column, value in filters.items()])


def _point_to_row(point: Any, id_field: str) -> dict[str, Any]:
    payload = point.payload if isinstance(point.payload, dict) else {}
    row = dict(payload)
    row[id_field] = point.id
    return row


def _candidate_ids(row_id: str) -> list[Any]:
    if row_id.lstrip("-").isdigit():
        return [int(row_id)]
    try:
        uuid.UUID(row_id)
    except ValueError:
        return []
    return [row_id]


PointTest = Callable[[Any], bool]


def id_prefix_test(prefix: str) -> PointTest:
    """Keep points whose id starts with a literal name prefix.

    Args:
        prefix (str): the literal prefix a leaf glob asked for.
    """

    def keep(point: Any) -> bool:
        return str(point.id).startswith(prefix)

    return keep


def value_prefix_test(column: str,
                      prefix: str,
                      basename: bool = False) -> PointTest:
    """Keep points whose payload value starts with a literal prefix.

    Args:
        column (str): the payload field the group level is named from.
        prefix (str): the literal prefix a group glob asked for.
        basename (bool): compare against the rendered path basename.
    """

    def keep(point: Any) -> bool:
        value = field_value(point.payload or {}, column)
        return value is not None and group_name(
            value, basename=basename).startswith(prefix)

    return keep


def exact_name_test(column: str, name: str, basename: bool,
                    seen: set[str]) -> PointTest:
    """Keep the first point of every raw value that renders as one name.

    Args:
        column (str): the payload field the group level is named from.
        name (str): the rendered segment a path spelled.
        basename (bool): compare against the rendered path basename.
        seen (set[str]): raw values already kept, shared across pages.
    """

    def keep(point: Any) -> bool:
        value = field_value(point.payload or {}, column)
        if value is None:
            return False
        raw = value_text(value)
        if raw in seen or group_name(raw, basename=basename) != name:
            return False
        seen.add(raw)
        return True

    return keep


async def _scroll_raw(client: Any,
                      collection: str,
                      flt: models.Filter | None,
                      limit: int,
                      keep: PointTest | None = None) -> list[Any]:
    # Without a test the limit bounds the scroll, which is the ordinary
    # capped listing. With one it bounds the MATCHES, because qdrant has
    # no prefix condition for a point id or a keyword field: the only
    # way to answer a glob for a row past the cap is to keep scrolling
    # and test each page here. A glob is a targeted request, so it pays
    # a scan of the collection where the plain listing pays one page.
    points: list[Any] = []
    offset: Any = None
    while len(points) < limit:
        batch, offset = await client.scroll(
            collection_name=collection,
            scroll_filter=flt,
            limit=SCROLL_BATCH if keep is not None else min(
                SCROLL_BATCH, limit - len(points)),
            offset=offset,
            with_payload=True,
            with_vectors=False,
        )
        points.extend(batch if keep is None else [p for p in batch if keep(p)])
        if offset is None:
            break
    return points[:limit]


def _is_index_required(exc: UnexpectedResponse) -> bool:
    content = exc.content
    text = content.decode() if isinstance(content, bytes) else str(content)
    return exc.status_code == 400 and "index required" in text.lower()


async def _ensure_indexes(client: Any, accessor: QdrantAccessor,
                          collection: str) -> None:
    if collection in accessor._indexes_ensured:
        return
    for field in accessor.config.group_by:
        await client.create_payload_index(
            collection_name=collection,
            field_name=field,
            field_schema="keyword",
        )
    accessor._indexes_ensured.add(collection)


async def _scroll_all(accessor: QdrantAccessor,
                      collection: str,
                      filters: dict[str, str],
                      limit: int,
                      keep: PointTest | None = None) -> list[Any]:
    client = await accessor.client()
    if not filters:
        return await _scroll_raw(client, collection, None, limit, keep)
    flt = _filter(filters)
    try:
        return await _scroll_raw(client, collection, flt, limit, keep)
    except UnexpectedResponse as exc:
        if not _is_index_required(exc):
            raise
    await _ensure_indexes(client, accessor, collection)
    return await _scroll_raw(client, collection, flt, limit, keep)


async def list_tables(accessor: QdrantAccessor) -> list[str]:
    client = await accessor.client()
    result = await client.get_collections()
    return sorted(item.name for item in result.collections)


async def table_exists(accessor: QdrantAccessor, name: str) -> bool:
    client = await accessor.client()
    return await client.collection_exists(name)


async def distinct_values(accessor: QdrantAccessor,
                          table: str,
                          column: str,
                          filters: dict[str, str],
                          limit: int,
                          prefix: str = "",
                          basename: bool = False) -> list[str]:
    keep = value_prefix_test(column, prefix, basename) if prefix else None
    points = await _scroll_all(accessor, table, filters, limit, keep)
    values = {
        value_text(value)
        for point in points
        if (value := field_value(point.payload or {}, column)) is not None
    }
    return sorted(values)


async def resolve_group(accessor: QdrantAccessor,
                        table: str,
                        column: str,
                        filters: dict[str, str],
                        name: str,
                        basename: bool = False) -> list[str]:
    """The raw payload values one rendered group segment stands for.

    A basename drops the value's parents, so two sources can render as
    the same directory. Telling them apart is a question about every
    point under the parent group, not about the first ``max_rows``: the
    scroll runs until it is exhausted or a second distinct value has
    rendered as ``name``, whichever comes first. One value is the
    answer; two is a collision for the caller to refuse.

    Args:
        accessor (QdrantAccessor): the mount's accessor.
        table (str): the collection.
        column (str): the payload field the group level is named from.
        filters (dict[str, str]): the parent groups, already resolved.
        name (str): the rendered segment a path spelled.
        basename (bool): whether the level renders basenames.
    """
    seen: set[str] = set()
    points = await _scroll_all(accessor, table, filters, 2,
                               exact_name_test(column, name, basename, seen))
    return sorted(
        value_text(field_value(point.payload or {}, column))
        for point in points)


async def rows_matching(accessor: QdrantAccessor,
                        table: str,
                        filters: dict[str, str],
                        limit: int,
                        prefix: str = "") -> list[dict[str, Any]]:
    keep: PointTest | None
    if prefix and accessor.config.name_field:

        def keep(point: Any) -> bool:
            return row_stem(_point_to_row(point, accessor.config.id_field),
                            accessor.config).startswith(prefix)
    else:
        keep = id_prefix_test(prefix) if prefix else None
    points = await _scroll_all(accessor, table, filters, limit, keep)
    return [_point_to_row(point, accessor.config.id_field) for point in points]


async def row_record(accessor: QdrantAccessor, table: str, id_field: str,
                     row_id: str) -> dict[str, Any] | None:
    ids = _candidate_ids(row_id)
    if not ids:
        return None
    client = await accessor.client()
    found = await client.retrieve(collection_name=table,
                                  ids=ids,
                                  with_payload=True,
                                  with_vectors=False)
    if not found:
        return None
    return _point_to_row(found[0], id_field)


async def search_rows(accessor: QdrantAccessor, table: str, query_text: str,
                      limit: int) -> list[dict[str, Any]]:
    key = (table, query_text, limit)
    cached = accessor.cached_search(key)
    if cached is not None:
        return cached
    client = await accessor.client()
    response = await client.query_points(
        collection_name=table,
        query=models.Document(text=query_text,
                              model=accessor.config.embedding_model),
        limit=limit,
        with_payload=True,
    )
    rows: list[dict[str, Any]] = []
    for point in response.points:
        row = _point_to_row(point, accessor.config.id_field)
        row["_score"] = point.score
        rows.append(row)
    accessor.store_search(key, rows)
    return rows
